proteuslib/property_models/tests/property_test_harness.py

```diff
@@ -0,0 +1,532 @@
+        for c_str in m.component_dict:
+            assert c_str in m.fs.properties.component_list
+            c = getattr(m.fs.properties, c_str)
+            c_type = m.component_dict[c_str]
+            assert isinstance(c, c_type)
+
+        # phases
+        assert isinstance(m.fs.properties.phase_list, Set)
+        assert len(m.fs.properties.phase_list) == len(m.phase_dict)
+        for p_str in m.phase_dict:
+            assert p_str in m.fs.properties.phase_list
+            p = getattr(m.fs.properties, p_str)
+            p_type = m.phase_dict[p_str]
+            assert isinstance(p, p_type)
+
+    @pytest.mark.unit
+    def test_parameters(self, frame):
+        m = frame
+
+        self.check_block_obj(m.fs.properties, m.param_obj_dict)
+        self.check_block_obj_coverage(m.fs.properties, m.param_obj_dict)
+
+        # test that the parameter variables are fixed
+        for v in m.fs.properties.component_objects(Var):
+            assert v.is_fixed()
+
+    @pytest.mark.unit
+    def test_default_scaling(self, frame):
+        m = frame
+
+        assert isinstance(m.fs.properties.default_scaling_factor, dict)
+        for t, sf in m.fs.properties.default_scaling_factor.items():
+            assert t in m.default_scaling_factor_dict.keys()
+            assert m.default_scaling_factor_dict[t] == sf
+
+    @pytest.mark.unit
+    def test_state_variables(self, frame):
+        m = frame
+        # state variables
+        state_vars_dict = m.fs.stream[0].define_state_vars()
+        assert len(state_vars_dict) == len(m.state_vars_list)
+        for v_str in state_vars_dict:
+            assert v_str in m.state_vars_list
+
+    @pytest.mark.unit
+    def test_metadata(self, frame):
+        m = frame
+        assert hasattr(m.fs.properties, 'define_metadata')
+        metadata = m.fs.properties.get_metadata().properties
+
+        # check method naming convention
+        for v_str in metadata.keys():
+            if v_str in m.state_vars_list:
+                assert metadata[v_str]['method'] is None
+            else:
+                assert metadata[v_str]['method'] == '_' + v_str
+
+    @pytest.mark.unit
+    def test_build(self, frame):
+        m = frame
+
+        # properties and state variables
+        prop_str_list = []
+        for obj_str, (obj_type, obj_units) in m.stateblock_obj_dict.items():
+            if obj_type is Expression or obj_type is Var:
+                prop_str_list.append(obj_str)
+
+        # check that the testing property list matches metadata
+        metadata = m.fs.properties.get_metadata().properties
+        assert len(metadata) == len(prop_str_list)
+        for v_str in metadata.keys():
+            assert v_str in prop_str_list
+
+        # check that properties are not built if not demanded
+        for obj_str in prop_str_list:
+            if obj_str not in m.state_vars_list:
+                assert not m.fs.stream[0].is_property_constructed(obj_str)
```

Comment:
This assumes that anything that is not a state variable is build-on-demand. This forces developers to use build on demand, and does not allow them to set certain properties as always present (e.g. enthalpy).

Reply:
Ok I can update this to only do it for variables and expressions that have an on demand method in the metadata

Reply:
Rather, I think you should have two separate tests - one for things that should always exist, and one for things that should be build-on-demand.